### ISSUE TITLE
Update to django 4.0

### ISSUE THREAD

**Issue by mrchrisadams** [thegreenwebfoundation/admin-portal#365]
Issue is for tracking the progress of updating the platform from Django 3.2 to Django 4.1

We have a number of dependencies to update before we can take the newer django to production

**Comment by mrchrisadams** [comment]
Updating from Django 3.2 to 4.0 

### Dependencies that needed to be updated

**django anymail**
`ImportError: cannot import name 'force_text' from 'django.utils.encoding'`
update to 8.6 LTS
https://anymail.dev/en/v9.0/changelog/

**django extensions**
Update to 3.2.1
`ImportError: cannot import name 'ugettext' from 'django.utils.translation'`
https://github.com/django-extensions/django-extensions/blob/main/CHANGELOG.md

**Comment by mrchrisadams** [comment]
Another one

**django_taggit_serializer**
`ImportError: cannot import name 'ugettext_lazy' from 'django.utils.translation'`

The previou maintainer hasn't merged in the PR
https://github.com/glemmaPaul/django-taggit-serializer/pull/48

So I'm trying this one:
https://github.com/adriangzz/dj-taggit-serializer

**Comment by mrchrisadams** [comment]
update any use of url() to `repath` in the urls file
`ImportError: cannot import name 'url' from 'django.conf.urls'`

Deprecated in 3.1

**Comment by mrchrisadams** [comment]
Ah nuts. We'll need to update the version of python to run django 4.0 onwards as they require python 3.9 at least.

**Comment by mrchrisadams** [comment]
I think we'll need to park this until we have a nicely documented way to choose the python we want to use.

Using pyenv on the server was a bit of a pain in the neck last time, and I didn't have much luck with asdf, which is what I use locally.

It might be the time to bite the bullet and Just Use Docker.

Until we've made a decision, I think we're stuck at Django 3.2, and I'm off on holiday tomorrow.

@tortila  - I'm tagging you in case you end up having a discussion with Hannah about this. Django 4.0 and 4.1 have much better control over the markup used by forms, but the python version is a blocker.

**Comment by tortila** [comment]
Thanks for tagging me Chris. Is Django version a blocker for Hannah? What's the priority of this issue?

I think there's a straightforward solution to switching to Python 3.9. We're using Ubuntu 20.04 and even though 3.8.5 is the default Python version shipped with the distro, `python3.9` is available in official package repository:
```
→ apt search python3.9
python3.9/focal-updates,focal-security,focal-updates,focal-security 3.9.5-3ubuntu0~20.04.1 amd64
  Interactive high-level object-oriented language (version 3.9)
```
We could also install a newer Python version if we added the [deadsnakes PPA](https://launchpad.net/~deadsnakes/+archive/ubuntu/ppa) on the production servers. 

I think that pyenv is great to use for development, but for production servers I would rather rely on a system package than build Python from sources.

**Comment by tortila** [comment]
It seems that Django 4.0 and 4.1 minimal requirement is Python 3.8: https://docs.djangoproject.com/en/4.1/releases/4.0/#python-compatibility

**Comment by hanopcan** [comment]
Hi @tortila, To complete the circle on this discussion as I know this has popped up in Trello card chat too. Having more control over the form html output would be ideal. I think its been mentioned that the current version only allows us to output forms as tables eg `form.as_table`

However, it looks as though we have found a way around these issues for now and I am able to style things in a way that is good enough.

So in short, no I don't believe this is a blocker for deployment of the new registration form 😄

**Comment by mrchrisadams** [comment]
We're no on django 5.x
